### PR TITLE
misc: optimize with precomputed constants to reduce computation

### PR DESCRIPTION
**Pull request by yenlinlai**
NewFromFloat is much more expensive than New or decimal copy. Reduce
    number of NewFromFloat by precomputing constant variables or by using
    New instead. However, some PI related NewFromFloat are still not
    removed as they are not all at same precision, force normalizing to
    a common PI will break tests.

### CODE REVIEW

**Comment by lantw44** [comment]
> misc: optimize with precomputed constants to reduce computation
> NewFromFloat is much more expensive than New or decimal copy. Reduce
> number of NewFromFloat by precomputing constant variable or

constant variables.

> by using using However,

What is 'by using using However'?

> some PI related NewFromFloat are still not removed as
> they are not all at same precision, force normalizing to a common PI
> will break tests.

**Comment by yenlinlai** [comment]
> > misc: optimize with precomputed constants to reduce computation
> > NewFromFloat is much more expensive than New or decimal copy. Reduce
> > number of NewFromFloat by precomputing constant variable or
> 
> constant variables.
> 
> > by using using However,
> 
> What is 'by using using However'?
> 
> > some PI related NewFromFloat are still not removed as
> > they are not all at same precision, force normalizing to a common PI
> > will break tests.

Fixed.

File: decimal.go [hunk 1]
```diff
@@ -1215,29 +1218,35 @@ func (d NullDecimal) MarshalJSON() ([]byte, error) {
 
 // Atan returns the arctangent, in radians, of d.
 func (d Decimal) Atan() Decimal {
-	if d.Equal(NewFromFloat(0.0)) {
+	if d.IsZero() {
 		return d
 	}
-	if d.GreaterThan(NewFromFloat(0.0)) {
+	if d.GreaterThan(Zero) {
```
Comment (lantw44) [inline decimal.go]: `d.IsPositive()`?

File: decimal.go [hunk 2]
```diff
@@ -1276,12 +1285,12 @@ func (d Decimal) Sin() Decimal {
...
 	M4PI := NewFromFloat(1.273239544735162542821171882678754627704620361328125) // 4/pi
 
-	if d.Equal(NewFromFloat(0.0)) {
+	if d.IsZero() {
 		return d
 	}
 	// make argument positive but save the sign
 	sign := false
-	if d.LessThan(NewFromFloat(0.0)) {
+	if d.Sign() < 0 {
```
Comment (lantw44) [inline decimal.go]: `d.IsNegative()`?

File: decimal.go [hunk 3]
```diff
@@ -1335,7 +1344,7 @@ func (d Decimal) Cos() Decimal {
 
 	// make argument positive
 	sign := false
-	if d.LessThan(NewFromFloat(0.0)) {
+	if d.Sign() < 0 {
```
Comment (lantw44) [inline decimal.go]: `d.IsNegative()`?

File: decimal.go [hunk 4]
```diff
@@ -1393,13 +1402,13 @@ func (d Decimal) Tan() Decimal {
...
 
-	if d.Equal(NewFromFloat(0.0)) {
+	if d.IsZero() {
 		return d
 	}
 
 	// make argument positive but save the sign
 	sign := false
-	if d.LessThan(NewFromFloat(0.0)) {
+	if d.Sign() < 0 {
```
Comment (lantw44) [inline decimal.go]: `d.IsNegative()`?

Comment (yenlinlai) [reply decimal.go]: Done

File: decimal.go (hunk 3 above)
Comment (yenlinlai) [reply decimal.go]: Done

File: decimal.go (hunk 2 above)
Comment (yenlinlai) [reply decimal.go]: Done

File: decimal.go (hunk 1 above)
Comment (yenlinlai) [reply decimal.go]: Done

**Review by lantw44 — APPROVED** [review]

**Review by jm-cobinhood — APPROVED** [review]

**Pull request merged**